graylog2-server/src/main/java/org/graylog/plugins/views/search/rest/SuggestionsResource.java

```diff
@@ -0,0 +1,114 @@
+import org.graylog.plugins.views.search.rest.suggestions.SuggestionEntryDTO;
+import org.graylog.plugins.views.search.rest.suggestions.SuggestionsDTO;
+import org.graylog.plugins.views.search.rest.suggestions.SuggestionsErrorDTO;
+import org.graylog.plugins.views.search.rest.suggestions.SuggestionsRequestDTO;
+import org.graylog2.audit.jersey.NoAuditEvent;
+import org.graylog2.plugin.indexer.searches.timeranges.InvalidRangeParametersException;
+import org.graylog2.plugin.indexer.searches.timeranges.RelativeRange;
+import org.graylog2.plugin.rest.PluginRestResource;
+import org.graylog2.shared.rest.resources.RestResource;
+import org.graylog2.shared.security.RestPermissions;
+
+import javax.inject.Inject;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@RequiresAuthentication
+@Api(value = "Search/Suggestions")
+@Path("/search/suggest")
+public class SuggestionsResource extends RestResource implements PluginRestResource {
+
+    private final PermittedStreams permittedStreams;
+    private final QueryEngine queryEngine;
+
+    @Inject
+    public SuggestionsResource(PermittedStreams permittedStreams, QueryEngine queryEngine) {
+        this.permittedStreams = permittedStreams;
+        this.queryEngine = queryEngine;
+    }
+
+    @POST
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation("Validate a search query")
+    @NoAuditEvent("Only validating query structure, not changing any data")
+    public SuggestionsDTO validateQuery(@ApiParam(name = "validationRequest") SuggestionsRequestDTO suggestionsRequest) {
+
+        final SuggestionRequest req = SuggestionRequest.builder()
+                .field(suggestionsRequest.field())
+                .input(suggestionsRequest.input())
+                .streams(adaptStreams(suggestionsRequest.streams()))
+                .timerange(Optional.ofNullable(suggestionsRequest.timerange()).orElse(defaultTimeRange()))
+                .build();
+
+        SuggestionResponse res = queryEngine.suggest(req);
+        final List<SuggestionEntryDTO> suggestions = res.suggestions().stream().map(s -> SuggestionEntryDTO.create(s.getValue(), s.getOccurrence())).collect(Collectors.toList());
+        final SuggestionsDTO.Builder suggestionsBuilder = SuggestionsDTO.builder(res.field(), res.input()).suggestions(suggestions);
+
+        res.suggestionError()
+                .map(e -> SuggestionsErrorDTO.create(e.type(), e.reason()))
+                .ifPresent(suggestionsBuilder::error);
+
+        return suggestionsBuilder.build();
+    }
+
+    private Set<String> adaptStreams(Set<String> streams) {
+        if (streams == null || streams.isEmpty()) {
+            return loadAllAllowedStreamsForUser();
+        } else {
+            // TODO: is it ok to filter out a stream that's not accessible or should we throw an exception?
```

Comment:
additionally: not sure if this filter should be done in the service?

Reply:
this will need some unified handling. I'll prepare a separate PR for the stream loading and filtering, as this is a repeated pattern that gets copy-pasted around.